### ISSUE TITLE
bug(forge): coverage does not work with return values assigned to library calls

### ISSUE THREAD

**Issue by PaulRBerg** [foundry-rs/foundry#4305]
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (249538f 2023-02-08T00:12:05.805004Z)

### What command(s) is the bug in?

forge coverage

### Operating System

macOS (Apple Silicon)

### Describe the bug

This bug has been first reported by @ZeroEkkusu in this [comment](https://github.com/foundry-rs/foundry/pull/3128#issuecomment-1241245086). I have opened a separate issue to track this because I think that it is an important and subtle bug, which can easily lead Foundry users astray (i.e. make them believe there's something wrong with their tests or library code when in fact there isn't).

The bug occurs when a return value is assigned to a library call, like this:

```solidity
library Helpers {
    function gte(uint256 element, uint256[] memory items) internal pure returns (bool) {
        if (items.length == 0) {
            revert("Items array is empty");
        }
        return element >= items[0];
    }
}

contract Foo {
    function gte(uint256 element, uint256[] memory items) external pure returns (bool) {
        // This is where the bug happens 👇
        return Helpers.gte(element, items);
    }
}
```

Given the code above, Forge will generate the following coverage report, which is incorrect:

<img width="980" alt="Screenshot 2023-02-08 at 12 22 30 PM" src="https://user-images.githubusercontent.com/8782666/217502768-2b9684b3-638d-41a2-8c25-1b20633bc2dc.png">

Assigning the library call to a variable fixes this:

```solidity
function gte(uint256 element, uint256[] memory items) external pure returns (bool) {
    bool result = Helpers.gte(element, items);
    return result;
}
```

However, this is bad practice. Production code should not be optimized for coverage reports.

As @onbjerg said [here](https://github.com/foundry-rs/foundry/pull/3128#issuecomment-1241924127), the bug may be related to the walking of the right-hand side of expressions.

**Comment by PaulRBerg** [comment]
@haythemsellami you can use the thumbs up emoji (👍) if you don't have anything to add.

**Comment by Picodes** [comment]
I opened a PR to fix this here https://github.com/foundry-rs/foundry/pull/5099. Locally the fix is working against my test suite.

**Comment by Evalir** [comment]
Closing, fixed with #5099
